src/core/net/tcp6_ext.cpp

```diff
@@ -0,0 +1,234 @@
+ *  1. Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *  2. Redistributions in binary form must reproduce the above copyright
+ *     notice, this list of conditions and the following disclaimer in the
+ *     documentation and/or other materials provided with the distribution.
+ *  3. Neither the name of the copyright holder nor the
+ *     names of its contributors may be used to endorse or promote products
+ *     derived from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ *  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ *  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ *  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ *  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @file
+ *   This file implements TCP/IPv6 socket extensions.
+ */
+
+#include "openthread-core-config.h"
+
+#if OPENTHREAD_CONFIG_TCP_ENABLE
+
+#include "tcp6_ext.hpp"
+
+#include "common/as_core_type.hpp"
+#include "common/code_utils.hpp"
+#include "common/error.hpp"
+#include "common/instance.hpp"
+#include "common/locator_getters.hpp"
+#include "common/log.hpp"
+#include "common/random.hpp"
+#include "net/checksum.hpp"
+#include "net/ip6.hpp"
+#include "net/netif.hpp"
+
+#include "../../third_party/tcplp/tcplp.h"
+
+namespace ot {
+namespace Ip6 {
+
+RegisterLogModule("TcpExt");
+
+void TcpCircularSendBuffer::Initialize(void *aDataBuffer, size_t aCapacity)
+{
+    this->mDataBuffer = static_cast<uint8_t *>(aDataBuffer);
+    this->mCapacity   = aCapacity;
+    ForceDiscardAll();
+}
+
+Error TcpCircularSendBuffer::Write(Tcp::Endpoint &aEndpoint,
+                                   const void *   aData,
+                                   size_t         aLength,
+                                   size_t &       aWritten,
+                                   uint32_t       aFlags)
+{
+    Error error = kErrorNone;
+
+    size_t bytesFree = FreeSpace();
+    size_t writeIndex;
+
+    uint32_t flags = 0;
+    size_t   bytesUntilWrap;
+
+    VerifyOrExit(aLength != 0);
+    VerifyOrExit(bytesFree != 0, aLength = 0);
+
+    /*
+     * Handle the case where we don't have enough space to accommodate all of the
+     * provided data.
+     */
+    if (aLength > bytesFree)
+    {
+        aLength = bytesFree;
+    }
```

Comment:
How about this to combine the checks?

```suggestion
    aLength = Min(aLength, bytesFree);
    VerifyOrExit(aLength != 0);
```

Reply:
I agree. That looks correct.